### ISSUE TITLE
LibProtocol: Suspiciously blocked due to fread()

### ISSUE THREAD

**Issue by awesomekling** [SerenityOS/serenity#5795]
I've seen the WebContent process deadlocked in LibProtocol a bunch of times now.

I'm not fully sure what's going on, but..

* LibProtocol's `Download::stream_into::(notifier->on_ready_to_read)` calls `fread()` (via `InputFileStream::read()` in response to `sys$select()` triggering a notifier.
* `fread()` may call `sys$read()` multiple times internally.
* If `sys$read()` reads everything currently available in the FIFO, we will block waiting for more data to come in.

So basically, the `on_ready_to_read` callback runs because there was something to read from the pipe. But then we keep trying to read even after exhausting the pipe.

I think maybe this happens when ProtocolServer is juggling multiple requests at the same time. Perhaps it's blocked waiting for space in another pipe that the LibProtocol side is supposed to drain.

cc @alimpfard @bugaevc 

Sorry about the sketchy details.. :shrug:

**Comment by alimpfard** [comment]
I'm not 100% sure on this (would need to take a look to be sure), but I'm pretty sure ProtocolServer will not block on any pipe
It should be writing whatever it can and moving on - at least, that's why I made it nonblocking...I think

**Comment by tomuta** [comment]
I believe there is a race condition where a non-blocking read can end up blocking unexpectedly, which prompted me to change how `Thread::ReadBlocker` and `Thread::WriteBlocker` are used and implemented in my smp pr. Unfortunately, that's not all done and require some very large changes, such as removing `Lock` from the FS/Net stack... Those blockers then essentially combine the `can_read/can_write` with blocking/instant unblocking without that race.

**Comment by Lubrsi** [comment]
This is almost guaranteed to happen when visiting certain pages like Google Doodles: https://www.google.com/doodles

**Comment by alimpfard** [comment]
@tomuta iirc the bug you're mentioning can only happen with multiple readers on the same fd, right?
I don't think that's the case here, there's only one reader :thinking:

**Comment by alimpfard** [comment]
@Lubrsi https://www.google.com/doodles doesn't deadlock for me, it _does_ say "did_receive refusing packet since buffer is full" a few times, but that's during (what I assume is) the page layout or CSS load.

`bt $(pidof ProtocolServer)` says that it's waiting on an event via `select()`, so it's not doing anything at all...

**Comment by Lubrsi** [comment]
Huh, weird. Sometimes it doesn't deadlock for me, but when it does it has the same issue (WebContent is stuck in fread, ProtocolServer is constantly in HTTP::Job::finish_up())

Stack for WebContent:
![Screenshot from 2021-03-14 22-57-57](https://user-images.githubusercontent.com/25595356/111087293-c2de7400-8518-11eb-84a4-f5f4f32a8b07.png)

perf_events for ProtocolServer: [perf_events.txt](https://github.com/SerenityOS/serenity/files/6137829/perf_events.txt)

**Comment by alimpfard** [comment]
So it looks like ProtocolServer is spinning while trying to write, and WebContent is stuck reading. 

Sounds like @awesomekling's `fread()` scenario is the only thing that can cause this, but then...if the download is done, why is the pipe still open? if it's not done, the next downloaded chunk will make `fread()` return (I think? or it might continue trying to read until it reads enough to fill the buffer?)

I'm not sure how this can be fixed...

**Comment by tomuta** [comment]
> @tomuta iirc the bug you're mentioning can only happen with multiple readers on the same fd, right?
> I don't think that's the case here, there's only one reader thinking

That's probably the most likely case, but IIRC it can also happen with just one consumer in some corner cases.

**Comment by bugaevc** [comment]
`fread()` always reads in a loop until it either reads the number of bytes requested, or hits an error (or EOF).

If you want to use it for non-blocking I/O, either request the size you somehow (e.g. `FIONREAD`) know is readily readable, or make the fd non-blocking (you'll then get `EWOULDBLOCK` errors from `fread`, but it should read as much data as possible anyway).

**Comment by ADKaster** [comment]
This sounds similar to what I was seeing in #5541 but that was with a select and a read. Which would point to it being the blocker issue.

**Comment by alimpfard** [comment]
@bugaevc `FIONREAD` would be nice (and actually fix a FIXME I have), but there's also the stdio buffer to consider - maybe that should expose an API that would return the max size that's certain not to block?

making it nonblocking will probably make normal downloads _really_ slow, and making it nonblocking after handing it to stdio sounds like begging for trouble...

**Comment by awesomekling** [comment]
> making it nonblocking will probably make normal downloads _really_ slow

Hmm, why is this? If you `read()` the fd until until you get `EAGAIN` every time the notifier wakes you up, where does the slowness come from?

**Comment by alimpfard** [comment]
it's just a guess, really, but doing so will force ProtocolServer to keep buffering downloaded chunks, and once that's done, the download rate will be limited to the pipe buffer limit?

:thinking: then again, it's normally also limited to that, so maybe not

**Comment by awesomekling** [comment]
Well right, downloads are already not able to progress until the client side reads from the 64 KB pipe buffer.

**Comment by awesomekling** [comment]
Okay I think I see what the real problem is: `sys$select` can unblock and return with false information about readable file descriptors.

If multiple file descriptors are readable *before* we block in `sys$select`, only the first readable file descriptor is actually marked as readable in the outgoing `fd_set`.

The blocker logic quite hard to follow, but basically, `SelectBlocker` aborts processing early and then `not_blocking()` doesn't `collect_unblocked_flags()` because we already set `m_did_unblock = true` in `SelectBlocker::unblock`.

The net result of this is that ProtocolServer has one or more TCP sockets that are disconnected/EOF and ready to read, but `sys$select` returns the wrong information and so ProtocolServer doesn't try to finish up the connection. Because we never read the EOF, we never EOF the pipe that `WebContent` is blocked on in `fread`, and we have our deadlock.

Does that make sense @tomuta?

**Comment by tomuta** [comment]
Hm yeah makes sense. Sounds like this condition here:
https://github.com/SerenityOS/serenity/blob/64cb04996f31c5163a3376a6ab649df59816430c/Kernel/ThreadBlockers.cpp#L388
should probably be
```c++
if (!m_should_block || !m_did_unblock)
```
